docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,462 @@
+      for height *h*, round *r*, as well as the proposed block the precommit messages refer to,
+    - node *m* uses the precommit messages to build a commit (not extended commit), which it then sends
+      to *all* validators in order to convince them they are late,
+    - all validators receive the *catch-up* message from *m*, decide on height *h*, and proceed to
+      height *h+1*.
+
+    At this point, *all* validators have advanced to *h+1* believing they are late, and so, expecting
+    the *hypothetical* leading majority of validators to propose for *h+1*. As a result, the
+    blockhain progress grinds to a halt.
+    A (rather complex) ad-hoc mechanism would need to be carried out by node operators to roll
+    back all validators to the precommit step of height *h*, round *r*.
+
+- **3.** *Require extended commits to be available at switching time*.
+
+    This is more involved than all previous solutions, and builds on an idea used in Solution 2:
+    vote extensions are actually not needed for Tendermint to make progress as long as the
+    validator is *certain* it is late.
+
+    We define two modes. The first is denoted *catch-up mode*, and Tendermint only calls
+    `FinalizeBlock` for each height when in this mode. The second is denoted *consensus mode*, in
+    which the validator considers itself up to date and fully participates in consensus and calls
+    `PrepareProposal`, `ExtendVote`, and `VerifyVoteExtension`, before calling `FinalizeBlock`.
+
+    The catch-up mode does not need vote extension information to make progress, as all it needs is the
+    decided block at each height to call `FinalizeBlock` and keep the state-machine replication making
+    progress. The consensus mode, on the other hand, does need vote extension information when
+    starting every height.
+
+    When a validator in consensus mode falls behind for whatever reason, e.g. cases (d), (e), (f),
+    and (g) above, we introduce the following key safety property:
+
+    - for every height *h*, a node refuses to switch to catch-up mode **until**:
+        - it has received and (light-client) verified all the blocks from *h+1* leading up to a
+          new height *h'*, where *h < h'*
+        - it has received an extended commit for *h'* and has verified:
+            - the precommit vote signatures in the extended commit
+            - the vote extension signatures in the extended commit: each is signed with the same
+              key as the precommit vote it extends
+
+    If the conditions above hold, namely receiving a valid sequence of blocks in the node's future,
+    starting at *h+1*, and an extended commit corresponding to the last block in the sequence, then
+    the node:
+
+    - switches to catch-up mode,
+    - applies all blocks (calling `FinalizeBlock` only), and
+    - switches back to consensus mode using the extended commit to propose in the rounds of
+      *h' + 1* where it is the proposer.
```

Comment:
I think this is a really nice idea and reminds me of the light block traces (imagine a light block with a long strand of headers whereby you skipping verify that light block and then backwards verify the hashes down to your trusted header and you've instantly batch verified a hundred headers - still under the light client security model). It also provides a somewhat adequate solution to the catching up problem i.e. how does a node know that it's at the head of the chain.

Reply:
Yes, the difference with pure light client verification is that in this case, unfortunately, we need all the blocks since we have to apply then so that the App state is consistent. However:
* We don't need to `PrepareProposal`/`ProcessProposal`, `ExtendVote`, and `VerifyVoteExtension`... but only `FinalizeBlock`, in A LOT of heights.
* In terms of verification, I think we can still benefit from the optimizations you mention... many blocks needing a minimal check (just last block hash probably)